File: frontend/gameapp-frontend/src/domain/value-objects/scenes/scene.js
import { SceneName } from './scene-name.js'
import { SceneDescription } from './scene-description.js'
import { Biome } from '../../enumerates/biome.js';

// Abstract base class Scene
export class Scene {
    constructor(name, description, biome) {
        if (!(name instanceof SceneName)) throw new TypeError("name must be SceneName");
        if (!(description instanceof SceneDescription)) throw new TypeError("description must be SceneDescription");
        if (!Object.values(Biome).includes(biome)) throw new TypeError("biome must be a valid Biome");

        this._name = name;
        this._description = description;
        this._biome = biome;
    }

    // getters
    get name() { return this._name; }
    get description() { return this._description; }
    get biome() { return this._biome; }
}

File: backend/GameApp.Domain/Entities/Scene.cs
using GameApp.Domain.Enumerates;
using GameApp.Domain.ValueObjects.Scenes;

namespace GameApp.Domain.Entities;

// Abtract entity class for a general scene
public abstract class Scene
{
    private readonly Guid Id;
    private readonly SceneName Name;
    private readonly SceneDescription Description;
    private readonly Biome Biome;

    // constructor EF
    protected Scene() { }

    // Default Constructor
    public Scene(SceneName name, SceneDescription description, Biome biome)
    {
        Id = Guid.NewGuid();

        Name = name;
        Description = description;
        Biome = biome;
    }

    // Restore Constructor
    public Scene(Guid id, SceneName name, SceneDescription description, Biome biome)
    {
        Id = id;
        Name = name;
        Description = description;
        Biome = biome;
    }

    // Getter
    public Guid GetGuid() => Id;
    public SceneName GetName() => Name;
    public SceneDescription GetDescription() => Description;
    public Biome GetBiome() => Biome;


}